### ISSUE TITLE
Q: col custom search + formatting

### ISSUE THREAD

**Issue by Mr-Tao** [code-and-effect/effective_datatables#80]
I’ve found this brilliant gem a few days ago.

I have a cell with a JSON serialized array. I need to list these items in a DataTable cell and I also need to search through them. As easy as it is to implement each of these requirements, I have spent the better part of the day trying to figure out how to achieve both.

- For the display part I did `col :participants, partial: 'participants/table_item', partial_as: 'event'` and put 
    ```HTML+ERB
    <ul>
      <% JSON::parse(event.participants).each do |p| %>
        <li><%= p %></li>
      <% end %>
    </ul>
    ``` 
    in the partial.
- For the search part I’ve set up full text search in PostgreSQL so that `Event.search_participants()` does the searching and I can write:
    ```ruby
    col :participants do |event|
      event.participants
    end.search do |collection, term, column, sql_column|
      collection = Event.search_participants(term)
    end
    ```
I would like to put these together, but I cannot as [You cannot use partial: … with the block syntax](https://github.com/code-and-effect/effective_datatables/blob/9bc38d0bf3d1a34d404286b30a7b2bdb4782fe5b/app/models/effective/effective_datatable/dsl/datatable.rb#L23)
So I tried

```ruby
col :participants do |event|
  tag.ul do
    JSON.parse(event.participants).each do |participant|
      tag.li participant
    end.search do |collection, term, column, sql_column|
      collection = Decision.search_participants(term)
    end
  end
end
```
but this render just an empty `<ul></ul>`. What am I doing wrong here?

Funnily I've skimmed through some old documentation too and it seems like overriding `search_column` method would suffice in the past.

**Comment by Mr-Tao** [comment]
Funy like you find a workaround just after writing the post…
```ruby
col :participants do |decision|
  tag.ul ActionController::Base.helpers.
      sanitize(JSON::parse(decision.participants).map { |i| tag.li i }.join )
end.search do |collection, term, column, sql_column|
  collection = Decision.search_participants(term)
end
```

There, however is another issue:
```ruby
col :category, visible: true, search: { as: :select, 
                                        collection: EventCategory.all, 
                                        multiple: true }
```
This does great job rendering multiple-choice selection, but items in the table are rendered as links to edit action. Passing `as: string` to col seems to work at first sight, but it breaks search since queries are formed like `SELECT  …  WHERE "event"."decision_category_id" ILIKE '%93%` instead `… IN (93)`.

**Comment by Mr-Tao** [comment]
Thanks, @matt-riemer, category indeed is a `belongs_to` object and `obj.category.to_s` did the job! `Event.joins` instigated the db to usurp all the memory instead :).

While scopes are working great, I am unable to get `filters` to perform.

I've set test filter like this:
```ruby
  filters do
    byebug
    filter :test_filter, "default-query"
  end
```
it looks like nothing I write into the filter field propagates and default gets always picked instead. After submitting any filter value I get:
```
    5:         def filter(name = nil, value = :_no_value, as: nil, label: nil, parse: nil, required: false, **input_html)
=>  6:           return datatable.filter if (name == nil && value == :_no_value) # This lets block methods call 'filter' and get the values
    7:
    8:           raise 'expected second argument to be a value' if value == :_no_value
    9:           raise 'parse must be a Proc' if parse.present? && !parse.kind_of?(Proc)
   10:
(byebug) name
:test_filter
(byebug) value
"default-query"
```

**Comment by matt-riemer** [comment]
Hmm, each filter just ends up as a form field in the filters section above the datatable.

If you just did `filter :start_date, Time.zone.now, required: true` does a date selector get displayed?  This should work just fine with your string too.

**Comment by Mr-Tao** [comment]
Filter fields are showing up as a plain text field. Is the date selector supposed to be a fancy picker?
![filter fields][filters]
The issue is that after submitting the form only default values are passed to query in the `collection` block.

[filters]: https://imgur.com/qWDqFoy.png "Filter fields as displayed above the DataTable"

**Comment by matt-riemer** [comment]
yea that date selector is supposed to be a fancy picker (if you include effective_form_inputs gem).

But it does look like your string filter is being displayed.  

Weird it wouldn't submit with whatever value you've typed into that search box!

Is there some kind of console error or otherwise going on here? What values do you see via the rails console when you click Apply?

**Comment by Mr-Tao** [comment]
I’ve included _effective_bootstrap_ – pretty fancy now :) Other than that the issue hold.

**Comment by matt-riemer** [comment]
Hey again @Mr-Tao,

I double checked and filters seem to be working just fine for me.... :)

effective_datatables 4.0.3
effective_resources 0.8.10
effective_bootstrap 0.0.16

I created a datatable with a string filter:

<img width="498" alt="rhythms_datatable_rb_ _tadum" src="https://user-images.githubusercontent.com/954468/40127804-79460a56-58ed-11e8-8544-5e0dac799251.png">

And it was displayed on my table:

![rhythms___tadum](https://user-images.githubusercontent.com/954468/40127848-91329fd0-58ed-11e8-99e8-accca596b2cb.png)

when I type in a value and click 'Apply', it posts to my server with the expected params[:filter]

<img width="1385" alt="1__ruby_" src="https://user-images.githubusercontent.com/954468/40127940-b5e2a2c6-58ed-11e8-9998-4cbd87c85674.png">

I hope this extra information can help you debug any problem!  Let me know

Cheers

**Comment by Mr-Tao** [comment]
Thanks, @matt-riemer. I’ve recreated your example with effective_(datatables|resources|bootstrap) and it works and the `filter` key is present in the `parameters` hash. Now I am trying to figure out why this key is missing when attempting to filter in my project.

**Comment by Mr-Tao** [comment]
Hi @matt-riemer, just a quick update. I did `rails new` and moved models, views and controllers to the new app (Rails 5.1.4 → 5.2). Good news is that filter key is present in the parameters hash now.

I somewhat suspect that all this has to do something with CSRF protection in Rails. I had to swap `protect_from_forgery with: :exception` to `skip_before_action :verify_authenticity_token` in the _application_controller.rb_ to render datatable in 5.2 in the first place. Otherwise the response to the Ajax POST was `422 (Unprocessable Entity)`.

Also there is an exception with EffectiveBootstrap in [base.js][rails.confirm]:
```
base.self-9c90a…1337.js?body=1:57 Uncaught TypeError: Cannot set property 'confirm' of undefined
    at base.self-9c90a…1337.js?body=1:57
    at base.self-9c90a…1337.js?body=1:70
```

[rails.confirm]: https://github.com/code-and-effect/effective_bootstrap/blob/65f946d3059d02bf183d56f94126df8cd126a859/app/assets/javascripts/effective_bootstrap/base.js.coffee#L39
